### ISSUE TITLE
Fully qualified name must be used inside a block when assigning to an ivar.

### ISSUE THREAD

**Issue by sol-vin** [crystal-lang/crystal#14827]
## Bug Report
It seems when assigning to an ivar and using a block at the same time any module references must use a FQN to be findable. The below minimal example illustrates the problem.

Thanks @Blacksmoke16 for the minimal example.
```crystal
module A
  struct B
  end

  struct C
    @data : B = 1.try { B.new }
  end
end

A::C.new
```
This produces an compile error `Error: undefined constant B`

### Temporary Fix

By changing `@data : B = 1.try { B.new }` to `@data : B = 1.try { A::B.new }` the program will compile again.

### Crystal
```
Crystal 1.13.1 [0cef61e]

LLVM: 18.1.1
Default target: x86_64-pc-windows-msvc
```

**Comment by HertzDevil** [comment]
This actually works with class variables because of this line:

https://github.com/crystal-lang/crystal/blob/670543afc0e29ee4be88c654af1818ef1e3302b0/src/compiler/crystal/semantic/class_vars_initializer_visitor.cr#L70

The same `pushing_type` is missing in the corresponding part of `Crystal::InstanceVarsInitializerVisitor`.

**Comment by cyangle** [comment]
It seems like jennifer fails to build due to the same issue here.

https://github.com/imdrasil/jennifer.cr/issues/460

**Comment by Blacksmoke16** [comment]
No, that failure is likely due to https://github.com/crystal-lang/crystal/issues/11181 given pointing at a method parameter.

**Comment by cyangle** [comment]
The jennifer failure is indeed caused by #11181, the tests passed after reverting the code change in #11208
